lib/remote/jsonrpcconnection.cpp

```diff
@@ -76,17 +82,50 @@ void JsonRpcConnection::HandleIncomingMessages(boost::asio::yield_context yc)
 		}
 
 		m_Seen = Utility::GetTime();
+		if (m_Endpoint) {
+			m_Endpoint->AddMessageReceived(jsonString.GetLength());
+		}
+
+		String rpcMethod("UNKNOWN");
+		ch::steady_clock::duration cpuBoundDuration(0);
+		auto start (ch::steady_clock::now());
 
 		try {
 			CpuBoundWork handleMessage (yc);
 
+			// Cache the elapsed time to acquire a CPU semaphore used to detect extremely heavy workloads.
+			cpuBoundDuration = ch::steady_clock::now() - start;
+
+			Dictionary::Ptr message = JsonRpc::DecodeMessage(jsonString);
+			if (String method = message->Get("method"); !method.IsEmpty()) {
+				rpcMethod = std::move(method);
+			}
+
 			MessageHandler(message);
 
 			l_TaskStats.InsertValue(Utility::GetTime(), 1);
+
+			auto total = ch::steady_clock::now() - start;
+
+			Log msg(total >= ch::seconds(5) ? LogWarning : LogDebug, "JsonRpcConnection");
+			msg << "Processed JSON-RPC '" << rpcMethod << "' message for identity '" << m_Identity
+				<< "' (took total " << toMilliseconds(total) << "ms";
+
+			if (cpuBoundDuration >= ch::seconds(1)) {
```

Comment:
@julianbrost Should a single message really take 5s to process before we warn about it? No wonder the non-debug logs didn't contain such messages.

Reply:
`total` also includes the time `CpuBoundWork` had to wait and a reason for this message was to be able to notice if the processing of JSON-RPC messages is delayed due to `CpuBoundWork` being overloaded.

What would you do differently? What would be an appropriate threshold? Would you want to log every message?

Reply:
I'd threshold both wait time and process time separately and PER log level. The higher the threshold, the higher the log level.

Reply:
How would you pick these thresholds? When would it log with a default configuration and what would it tell us?

Reply:
I'd discuss this in person with all core devs.

But now as you said the very latter, the better question is: even if we hit enough customer's messages, which poor pig of us shall aggregate those logs? /s I.e, shouldn't we aggregate the stuff directly (https://github.com/Icinga/icinga2/pull/10266#discussion_r1878420487)?

Reply:
That log message isn't really for aggregation, so the answer to your sarcastic question is nobody. If you see that log message, it can tell you one of two things:
1. There is a single JSON-RPC that blocks the connection for very long (less likely, more to actually confirm that this doesn't happen)
2. Shit has hit the `CpuBoundWork` fan and that delays everything (more likely)